### PR TITLE
migrate dolt pull to use sql queries

### PR DESCRIPTION
**Pull request by stephkyou**
This change updates `dolt pull` to use the appropriate sql engine to generate results.

Related: https://github.com/dolthub/dolt/issues/3922

### CODE REVIEW

**Comment by macneale4** [comment]
output status of `dolt merge` changing so adding the minor version bump label

File: integration-tests/bats/pull.bats
Comment (macneale4) [inline integration-tests/bats/pull.bats:39]: the CD step isn't required

File: integration-tests/bats/pull.bats
Comment (macneale4) [inline integration-tests/bats/pull.bats:49]: My feedback applies to pretty much every test here - there isn't any validation that the HEAD actually changed as a result of the merge. It's not sufficient to look at the output text and status. Is there a common pattern you could apply to all tests which verifies that the content of each parent in the merge is in the final result?

**Review by macneale4 — CHANGES_REQUESTED** [review]
Excellent so far. A few small nits and questions. Main thing to fix is verifying results in bats tests.

File: go/cmd/dolt/commands/pull.go
Comment (stephkyou) [reply go/cmd/dolt/commands/pull.go:122]: Staged changes in merges with `--no-commit` won't stick otherwise. I have a clarifying comment in merge that I'll put here too.

File: go/cmd/dolt/commands/pull.go
Comment (stephkyou) [reply go/cmd/dolt/commands/pull.go:138]: Yes, it's to make the error non-nil for `HandleVErrAndExitCode`. It doesn't seem ideal but it does work :/

File: integration-tests/bats/pull.bats
Comment (macneale4) [inline integration-tests/bats/pull.bats:213]: fully spell it out: "--force", I did a search to find it and was confused when it wasn't found

File: integration-tests/bats/sql-pull.bats
Comment (macneale4) [inline integration-tests/bats/sql-pull.bats:164]: These tests probably need a little explanation as to why you are adding FK. My read is that you are disabling them so you can get the contraint violations on merge, which is great. But it takes a little puzzling out for the reader to determine that.

**Review by macneale4 — CHANGES_REQUESTED** [review]
Code looks fine, I left a couple small comments. If the tests were passing I'd say ship it, but I see ~80 test failures. So I'll wait for you to sort those out then give one more look.

File: integration-tests/bats/remotes.bats
Comment (macneale4) [inline integration-tests/bats/remotes.bats:1352]: Is it the case that config wasn't required before? This is surprising.

File: go/cmd/dolt/commands/pull.go
Comment (macneale4) [reply go/cmd/dolt/commands/pull.go:144]: We talked about this in person. Let's just abort if we can't get the HEAD. Can always improve in the future

**Review by macneale4 — CHANGES_REQUESTED** [review]
A few nits, but very close!

File: go/cmd/dolt/commands/merge.go
Comment (stephkyou) [reply go/cmd/dolt/commands/merge.go:156]: Oops, was leftover from testing, will remove.

File: go/cmd/dolt/commands/merge.go
Comment (stephkyou) [reply go/cmd/dolt/commands/merge.go:311]: `dolt_merge` returns `hash`, `fast_forward`, `conflicts`. `dolt_pull` returns `fast_forward`, `conflicts`.

File: go/cmd/dolt/commands/pull.go
Comment (stephkyou) [reply go/cmd/dolt/commands/pull.go:267]: There is a test that checks that an empty remote name does not panic. I guess this makes more sense in the stored procedure context. I'll remove the test from the CLI bats tests and remove this logic.

File: integration-tests/bats/conflict-cat.bats
Comment (stephkyou) [reply integration-tests/bats/conflict-cat.bats:121]: I don't know, it seems like it should be. The `dolt_schema_conflicts` table is empty after the merge which is what determines the message. Perhaps the content violation takes precedent over the schema conflict?

File: integration-tests/bats/conflict-cat.bats
Comment (stephkyou) [reply integration-tests/bats/conflict-cat.bats:146]: Seems like same situation as above.

File: integration-tests/bats/remotes.bats
Comment (stephkyou) [reply integration-tests/bats/remotes.bats:1352]: It seems that config wasn't required to do a fast-forward pull before, but merge always requires a config so I think it makes sense for pull to do the same. It's also hard to allow fast-forward pulls without a config because that logic would have to be taken care of in the sql context.

File: go/cmd/dolt/commands/merge.go
Comment (macneale4) [reply go/cmd/dolt/commands/merge.go:309]: add a comment in the code to that effect. that was lost on me

File: go/cmd/dolt/commands/pull.go
Comment (macneale4) [reply go/cmd/dolt/commands/pull.go]: yeah, its shouldn't panic, this is true, but let's just print an error or something rather than silently consuming it

File: integration-tests/bats/remotes.bats
Comment (macneale4) [reply integration-tests/bats/remotes.bats:1352]: Agreed. Let's alway require it.

File: go/cmd/dolt/commands/push.go
Comment (macneale4) [inline go/cmd/dolt/commands/push.go:202]: I think there is  still a need for the error, yes? If there are no rows then you'll get an index out of bounds error here.

**Review by macneale4 — APPROVED** [review]
One small thing in your latest change to tweak.

Get a sanity check from Jason about the merge conflict errors, then you are welcome to ship without another review from me!

**Pull request merged**
